morpho/processors/diagnostics/CalibrationProcessor.py

```diff
@@ -0,0 +1,278 @@
+            if self.quantile == True:
+                for param_name in calib_bounds:
+                    calib_bounds[param_name].append(self._get_quantile_bounds(posterior_arrays[param_name]))
+            else:
+                for param_name in calib_bounds:
+                    bounds = self._get_highest_density_bounds(posterior_arrays[param_name], alpha)
+                    calib_bounds[param_name].append(bounds)
+                    #Consistency-with-zero checks are only possible for HDIs
+                    if bounds[0] == 0.0:
+                        consistent_with_zero[param_name] += 1
+            
+            #Tracking and optionally printing information about the intervals
+            bs={key:val[i] for key, val in calib_bounds.items()}
+            if self.verbose==True:
+                logger.info('\n---------------------EXPERIMENT #{}:---------------------'.format(i))
+            self._report_post_param_info(input_vals, posterior_arrays, bs, sums)
+            if self.verbose==True:
+                logger.info('\n--------------------------------------------------------')
+                
+            #Determining whether intervals contain inputted values
+            for param_name in calib_bounds:
+                if len(self.cred_interval) == 1:
+                    if input_vals[param_name] <= bs[param_name][0]:
+                        n_recovered_inputs[param_name] += 1
+                elif len(self.cred_interval) == 2:
+                    if bs[param_name][0] <= input_vals[param_name] <= bs[param_name][1]:
+                        n_recovered_inputs[param_name] += 1
+            
+        #Calculating an interval coverage for each parameter
+        for run in self.failed_runs:
+            self.files.remove(run)
+        coverages = {p:n_recovered_inputs[p]/float(len(self.files)) for p in self.in_param_names}
+        
+        #Reporting calibration results
+        self._report_calibration_results(calib_bounds, consistent_with_zero, coverages, sums)
+        
+        return coverages
+        
+
+    def _load_inputs_and_posteriors(self, filename):
+        """
+        For a given root file, returns two dictionaries, one containing inputted values and the other containing posterior arrays.
+        """
+        if self.verbose:
+            logger.info("Reading input values and posterior arrays")
+        in_reader_config = {
+            "action": "read",
+            "tree_name": self.root_in_tree,
+            "filename": filename,
+            "variables": self.in_param_names
+        }
+        post_reader_config = {
+            "action": "read",
+            "tree_name": self.root_post_tree,
+            "filename": filename,
+            "variables": self.post_param_names
+        }
+        rin, rpost = IOROOTProcessor("reader"), IOROOTProcessor("reader2")
+        rin.Configure(in_reader_config)
+        rpost.Configure(post_reader_config)
+        rin.Run()
```

Comment:
here you should try and catch potential errors that could happen during the processor.Run().
If the processor fails, it returns False, so simply catching this would be fine:
```python
if not run.Run():
    logger.error("Failed to run reader processor")
    raise
```
Maybe you can make this exception caught by the except line 85: for that you can add the right exception (maybe interrupted error) using `raise InterruptedError`  and catch it line 85 with an additional `except InterruptedError as error` (either stopping the processor or doing something else...)

Reply:
Same for rpost... :)

Reply:
Added a 
```python
            except RuntimeError as error:
                logger.warning("Caught processor error; passing...")
                pass
```
in `perform_calibration()` and 
```python
        if not rin.Configure(in_reader_config):
            logger.warning("Failed to Configure <{}> processor".format(rin.name))
            raise RuntimeError
        if not rpost.Configure(post_reader_config):
            logger.warning("Failed to Configure <{}> processor".format(rpost.name))
            raise RuntimeError
        if not rin.Run():
            logger.warning("Failed to Run <{}> processor".format(rin.name))
            raise RuntimeError
        if not rpost.Run():
            logger.warning("Failed to run <{}> processor".format(rpost.name))
            raise RuntimeError
```
in the load_inputs

This should cover it for the time being; maybe one day we will have a proper exception system in morpho...

Reply:
Looks great. I'll keep the need for an updated exception system in mind.